vm/qemu/qmp.go

```diff
@@ -0,0 +1,108 @@
+package qemu
+
+import (
+	"fmt"
+	"net"
+	"encoding/json"
+	"errors"
+)
+type qmpVersion struct {
+	Package string
+	QEMU    struct {
+		Major int
+		Micro int
+		Minor int
+	}
+}
+
+type qmpBanner struct {
+        QMP struct {
+		Version qmpVersion
+        }
+}
+
+type qmpCommand struct {
+	Execute   string      `json:"execute"`
+	Arguments interface{} `json:"arguments,omitempty"`
+}
+
+type hmpCommand struct {
+	Command string `json:"command-line"`
+	CPU     int    `json:"cpu-index"`
+}
+
+type qmpResponse struct {
+	Error string       `json:"error,omitempty"`
+	Return interface{} `json:"return,omitempty"`
+}
+
+func (inst *instance) qmp_check_connection() error {
+	if inst.mon != nil {
+		return nil
+	}
+
+	addr := fmt.Sprintf("127.0.0.1:%v", inst.monport)
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return err
+	}
+
+	dec := json.NewDecoder(conn)
```

Comment:
There is a caveat for NewDecoder/NewEncoder -- they don't do internal buffering, like most Go's io.Reader/Writer writers because there is a bufio package that can do this if necessary.
So if you do NewDecoder on top of, say, bytes.Buffer, then it's the right thing to do it directly. bytes.Buffer provides efficient small read/writes.
But if you do it on top of net.Conn, then the right thing is:
```
dec := json.NewDecoder(bufio.NewReader(conn))
```
If you think about it, there is not way to do efficient buffered reads without potentially over-reading data (consuming part of the next message). Since you don't have this issue, it has to be issuing lots of small inefficient read syscalls.

So, for encoding I would simply use json.Marshal as you do in qmp. It's simpler and avoids these issues.

Decoding I would do:
```
dec := json.NewDecoder(bufio.NewReader(conn))
```
but then we have to store it in inst and reuse. If we drop and recreate it, it will overread due to buffering.

Reply:
Ah rats. This is wrong PR you are reviewing, I've made it against my `master` branch, it should have been `ppc64`. Sorry about that.

Reply:
`The decoder introduces its own buffering` says https://golang.org/pkg/encoding/json/#Decoder.Decode
Or this is something different?

Reply:
You are right: "The decoder introduces its own buffering and may read data from r beyond the JSON values requested". But then your code is broken :) as you throw away decoded with buffered data...